### ISSUE TITLE
IO Feature Request: A way to deduce the file/path that a fileReader or fileWriter is open on

### ISSUE THREAD

**Issue by psath** [chapel-lang/chapel#23066]
The use case that I hit upon for the above is capturing the file path from within the `readThis` or `writeThis` methods, which only have access to a `fileReader` or `fileWriter` respectively.

If an API function can be added for both of those that returns a `file` object, then you should just be able to use its `.path` method to get the path as a string.

I'm still pretty green to using Chapel's I/O for plain-old-binary-files, much less in a parallel context or more exotic file types. If there are any safety considerations to the above, that'd be good to know and reflect on how to refactor my code pattern.

**Comment by mppf** [comment]
As far as I know, it would be safe to get a `file` out of a `fileReader` or `fileWriter`. I think the lack of it is an oversight. In terms of next step, I think it's just settling on a name for it. The implementation can just run `qio_channel_get_file_ptr` which already exists. (And, you could call that function for now, as a workaround).

If you wanted to read some stuff in parallel, you could get the `file` and then create additional `fileReader`s from it, without needing the path at all.

The only thing to be careful of is that each `fileReader` / `fileWriter` will do separate buffering. That means
 * if you are dividing up a file to write different pieces in parallel, make sure to specify disjoint regions when creating the `fileWriter`
 * if you are doing reads and writes to the same area of a file, you will have to manually flush to see the writes in the reads

**Comment by psath** [comment]
Thanks for the tips, still need the `file` but I can look at `qio_channel_get_file_ptr`.

Thankfully my inFile and outFile are separate, so the overlapping embarrassingly parallel reads will not conflict with the disjoint parallel writes and I won't have to worry about flushing between writes and reads, but good to know for future work.

**Comment by mppf** [comment]
FWIW here are some parts towards a PR for this

<details>

a patch that adds `.getFile()` - name needs more discussion

```
diff --git a/modules/standard/IO.chpl b/modules/standard/IO.chpl
index 9f548a905a..9983951d42 100644
--- a/modules/standard/IO.chpl
+++ b/modules/standard/IO.chpl
@@ -3771,6 +3771,35 @@ proc fileReader._getFp(): (bool, c_ptr(c_FILE)) {
   }
 }
 
+/* Returns the :record:`file` for a :record:`fileReader` */
+proc fileReader.getFile() {
+  // get the file pointer
+  var f: qio_file_ptr_t;
+  qio_channel_get_file_ptr(this._channel_internal, f);
+  // bump the reference count
+  qio_file_retain(f);
+  // store it in a record file
+  var ret: file;
+  ret._home = this._home;
+  ret._file_internal = f;
+  return ret;
+}
+
+/* Returns the :record:`file` for a :record:`fileWriter` */
+proc fileWriter.getFile() {
+  // get the file pointer
+  var f: qio_file_ptr_t;
+  qio_channel_get_file_ptr(this._channel_internal, f);
+  // bump the reference count
+  qio_file_retain(f);
+  // store it in a record file
+  var ret: file;
+  ret._home = this._home;
+  ret._file_internal = f;
+  return ret;
+}
+
+
 /*
 
 Represents a Unicode codepoint. I/O routines (such as :proc:`fileReader.read`
```

a test of it

``` chapel
use IO;

var f2: file;

{
  var f = openTempFile();

  {
    var writer = f.writer();
    writer.writeln("hello");

    assert(f == writer.getFile());
    f2 = writer.getFile();
  }

  {
    var reader = f.reader();

    var s: string;
    f.reader().read(s);
    assert(s == "hello");
    assert(f == reader.getFile());
  }

  {
    f2.writer().writeln("byebye");
    var s: string;
    f2.reader().read(s);
    assert(s == "byebye");
  }
}

{
  var s: string;
  f2.reader().read(s);
  assert(s == "byebye");
}
```

</details>

**Comment by bradcray** [comment]
> FWIW here are some parts towards a PR for this

@mppf: Given that I'm only following this very loosely, what are the other parts that are missing?  (I see the item about deciding what the name should be).

**Comment by mppf** [comment]
> what are the other parts that are missing?

Main thing is choose a name / strategy for stabilization. E.g. it could proceed with what I have here but be unstable for now.
  * Note that, initially, I wanted to use `myWriter.file` as the naming strategy, but this would require adjusting various fileReader / fileWriter methods to use `IO.file` instead of `file` in their signatures. I balked at that but maybe it would be OK.

Other than that it's to create a PR & get review & run testing etc.

I didn't want to handle either of these myself right now, which is why I posted a patch here.
